tools/mason/PkgConfig.chpl

```diff
@@ -0,0 +1,156 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+use Path;
+use MasonUtils;
+use MasonHelp;
+
+
+/* Checks to see that pkg-config is installed */
+proc pkgConfigExists() throws {
+  var status = runWithStatus("pkg-config --version", false);
+  if status != 0 {
+    throw new MasonError("pkg-config is not installed");
+  }
+  return true;
+}
+
+/* Given arguments from the command line
+   prints .pc file for a pkg-config package */
+proc masonPkgPcLookup(args) {
+  try! {
+    if args.size < 3 || args[2] == "--help" || args[2] == "-h" {
+      masonPkgPcHelp();
+    }
+    else {
+      const pkgName = args[2];
+      if pkgConfigExists() {
+        if pkgExists(pkgName) {
+          printPkgPc(pkgName);
+        }
+        else {
+          writeln("Mason could not find: " + pkgName);
+          writeln("Make sure the package is installed and it's .pc file is in the PKG_CONFIG_PATH");
+        }
+      }
+    }
+  }
+  catch e: MasonError {
+    writeln(e.message());
+  }
+}
+
+/* Prints a pc for user debugging */
+proc printPkgPc(pkgName: string) throws {
+  try! {
+    var pcDir = "".join(getPkgVariable(pkgName, "--variable=pcfiledir")).strip();
+    var pcFile = joinPath(pcDir, pkgName + ".pc");
+    var pc = open(pcFile, iomode.r);
+    for line in pc.lines() {
+      write(line);
+    }
+  }
+  catch e: FileNotFoundError {
+    // start issue about this
+    //throw new MasonError("Package exists but Mason could not find it's .pc file");
+    writeln("Package exists but Mason could not find it's .pc file");
+  }
+}
+
+
+/* Gets a single variable from pkg-config
+   given pkgname and variable */
+proc getPkgVariable(pkgName: string, option: string) {
+
+  var command = " ".join("pkg-config", pkgName, option);
+
+  var lines: [1..0] string;
+  var cmd = command.split();
+  var sub = spawn(cmd, stdout=PIPE);
+  sub.wait();
+
+  var line:string;
+  for line in sub.stdout.lines() {
+    if line.length > 1 then
+    lines.push_back(line);
+  }
+  
+  return lines;
+ }
+
+/* Queries pkg-config for package existance */
+proc pkgExists(pkgName: string) : bool {
+
+  const debug = " --print-errors";
+  var command = "pkg-config " + pkgName + " --exists";
+  const status = runWithStatus(command);
+  if status != 0 {
+    return false;
+  }
+  return true;
+}
+
+/* Retrieves build information for MasonUpdate */
+proc getPkgInfo(pkgName: string, version: string) throws {
+
+  var pkgDom: domain(string);
+  var pkgToml: [pkgDom] unmanaged Toml;
+  var pkgInfo: unmanaged Toml = pkgToml;
+
+  if pkgExists(pkgName) {
+    const pcVersion = "".join(getPkgVariable(pkgName, "--modversion")).strip();
+    const libs = "".join(getPkgVariable(pkgName, "--libs")).strip();
+    const include = "".join(getPkgVariable(pkgName, "--variable=includedir")).strip();
```

Comment:
Would `--cflags` be more robust instead instead of specifying the `includedir`? @mppf might have a good idea.

Reply:
I'd expect we'd want the simplest use of pkg-config possible, where we ask for `--libs` (which really means *linker flags*) and `--cflags` which is the compilation flags (including `-I` arguments).

Is there a particular reason we shouldn't do that?

Reply:
pkg-config assumes a number of things about the system including that if a package's include dir is `/usr/include/` then it prints nothing because that location is assumed by the c compiler. This is a way to ensure that the correct information is always put into the lock file.

Reply:
Why do we need to put /usr/include into the lock file? What if the pkgconfig package specifies another C flag (e.g. -shared etc)

Reply:
`/usr/include` needs to be included because we can't special case when then path is `/usr/include` during build because if nothing is returned than the toml file is invalid. mason build expects there to be something there and if it is not there then the toml file is invalid and it cant read the toml file anyway. Now we could special case mason update and put a placeholder into the toml file but in my opinion, at that point, we should just put `usr/include` in the lock file.

Reply:
If you just need a placeholder, "" could work couldn't it? Or something along those lines?

It seems really weird to me to query pkg-config so that we get a placeholder that our TOML parser can parse. That seems like something we should handle in a different way. It also seems weird to put `/usr/include` into the lock file, since as mentioned above, the C compiler will search there for includes anyway.